docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -294,254 +296,289 @@ TODO
+`InfoResponse`), then `defaultLane` will be set to 1, and `lanes` will have only one entry for the
+default lane. In this case, the new mempool's behaviour will be equivalent to that of the current mempool.
+
+`CListMempool` also contains the cache, which is only needed before transactions have a lane
+assigned. Since the cache is independent of the lanes, we do not need to modify it.
+
+### Configuration
+
+For an MVP, all lanes can share the same mempool configuration. In this scenario, all lanes will be
+capped by the total mempool capacities as currently defined in the configuration. Namely, these are:
+- `Size`, the total number of transactions allowed in the mempool, 
+- `MaxTxsBytes`, the maximum total number of bytes of the mempool, and
+- `MaxTxBytes`, the maximum size in bytes of a single transaction accepted into the mempool.
+
+With lanes, the total size of the mempool will be the sum of the sizes of all lanes.
+
+Additionally, the `Recheck` and `Broadcast` flags will apply to all lanes or to none. Remember that,
+if `PrepareProposal`'s app logic can ever add a new transaction, it becomes _always_ mandatory to recheck remaining transactions in the
+mempool, so there is no point in disabling `Recheck` per lane.
+
+### Adding transactions to the mempool
+
+When validating a transaction received for the first time with `CheckTx`, the application will
+optionally return its lane in the response.
+```protobuf
+message CheckTxResponse {
+  ...
+  uint32 lane = 12;
+}
+```
+The callback that handles the first-time CheckTx response will append the new mempool entry to the
+corresponding `CList`, namely `lanes[lane]`, and update the other auxiliary variables accordingly.
+If `lane` is 0, it means that the application did not set any lane in the response message, so the
+transaction will be assigned to the default lane.
+
+### Removing transactions from the mempool
+
+A transaction may be removed in two scenarios: when updating a list of committed transactions, and
+during rechecking if the transaction is reassessed as invalid. In either case, the first step is to
+identify the lane the transaction belongs to by accessing the `txLanes` map. Then, we remove the
+entry from the CList corresponding to its lane and update the auxiliary variables accordingly.
+
+Since the broadcast goroutines are constantly reading the list of transactions to disseminate them,
+it's important to prioritize the removal of transactions from high-priority lanes first.
+
+When updating the mempool, there is potential for a slight optimization by removing transactions
+from different lanes in parallel. To achieve this, we would first need to preprocess the list of
+transactions to determine the lane of each transaction. However, this optimization has minimal
+impact if the committed block contains few transactions. Therefore, we decided to exclude it from
+the MVP.
+
+### Reaping transactions for creating blocks
```

Comment:
Any discussion regarding starvation?

Namely, no guarantee that a tx in a low priority lane will eventually be included in the output?

I mentioned that to you guys at some point, and I still think that the way the `p2p/MConnection` does that is probably more adequate. I think that considering this approach should at least be discussed here, as it is an ADR.

Reply:
It was not mentioned so I added it. In essence it's the same as for tx dissemination. I agree we should discuss what P2P does and compare it to the proposed scheduling algorithm.

Reply:
I added a section Prioritization logic that includes the current scheduling algorithm in the P2P reactor.